packages/react/src/components/FocusTrapZone/FocusTrapZone.tsx

```diff
@@ -97,284 +92,230 @@ export const FocusTrapZone: React.FunctionComponent<IFocusTrapZoneProps> & {
-      internalState.disposeClickHandler = undefined;
+  // Update window event handlers when relevant props change
+  React.useEffect(() => {
+    const disposables: Array<() => void> = [];
+
+    if (forceFocusInsideTrap) {
+      disposables.push(on(window, 'focus', forceFocusOrClickInTrap, true));
+    }
+    if (!isClickableOutsideFocusTrap) {
+      disposables.push(on(window, 'click', forceFocusOrClickInTrap, true));
     }
-  }, [forceClickInTrap, forceFocusInTrap, forceFocusInsideTrap, isClickableOutsideFocusTrap, internalState]);
 
-  // Updates eventHandlers and cleans up focusStack when the component unmounts.
-  React.useEffect(() => {
-    const parentRoot = root.current;
-    updateEventHandlers();
     return () => {
-      // don't handle return focus unless forceFocusInsideTrap is true or focus is still within FocusTrapZone
-      if (!disabled || forceFocusInsideTrap || !elementContains(parentRoot, doc?.activeElement as HTMLElement)) {
-        returnFocusToInitiator();
-      }
+      disposables.forEach(dispose => dispose());
     };
-    // eslint-disable-next-line react-hooks/exhaustive-deps -- Should only run on mount.
-  }, [updateEventHandlers]);
+    // eslint-disable-next-line react-hooks/exhaustive-deps -- should only run when these two props change
+  }, [forceFocusInsideTrap, isClickableOutsideFocusTrap]);
 
-  // Updates focusStack and the previouslyFocusedElementOutsideTrapZone on prop change.
+  // On prop change or first render, focus the FTZ and update focusStack if appropriate
   React.useEffect(() => {
-    const newForceFocusInsideTrap = forceFocusInsideTrap !== undefined ? forceFocusInsideTrap : true;
-    const newDisabled = disabled !== undefined ? disabled : false;
+    // Do nothing if disabled, or if it's a re-render and forceFocusInsideTrap is false
+    // (to match existing behavior, the FTZ handles first focus even if forceFocusInsideTrap
+    // is false, though it's debatable whether it should do this)
+    if (disabled || (!isFirstRender && !forceFocusInsideTrap) || !root.current) {
+      return;
+    }
 
-    // Transition from forceFocusInsideTrap / FTZ disabled to enabled.
-    if (!newDisabled || newForceFocusInsideTrap) {
-      if (currentDisabledValue) {
-        return;
-      }
+    // Transition from forceFocusInsideTrap / FTZ disabled to enabled (or initial mount)
+    FocusTrapZone.focusStack!.push(internalState.focusStackId);
 
-      FocusTrapZone.focusStack.push(id);
+    const elementToFocusOnDismiss = props.elementToFocusOnDismiss || (doc!.activeElement as HTMLElement | null);
 
-      internalState.previouslyFocusedElementOutsideTrapZone = elementToFocusOnDismiss
-        ? elementToFocusOnDismiss
-        : (doc!.activeElement as HTMLElement);
-      if (!disableFirstFocus && !elementContains(root.current, internalState.previouslyFocusedElementOutsideTrapZone)) {
-        focus();
-      }
-      if (!internalState.unmodalize && root.current && enableAriaHiddenSiblings) {
```

Comment:
Modalization (and cleanup) is now handled in a separate effect, because the conditions where it needs to change aren't quite the same as the conditions where we need to trap or restore focus.